test/hotspot/jtreg/runtime/ClassInitErrors/InitExceptionUnloadTest.java

```diff
@@ -122,10 +124,26 @@ static void test() throws Throwable {
             }
         }
         cl = null;
-        ClassUnloadCommon.triggerUnloading();  // should unload these classes
-        for (String className : classNames) {
-          ClassUnloadCommon.failIf(wb.isClassAlive(className), "should be unloaded");
+
+        Set<String> loadedClasses = new HashSet<>(Set.of(classNames));
+        int attempt = 0;
+        while (!loadedClasses.isEmpty() && attempt < 20) {
+            ClassUnloadCommon.triggerUnloading();
```

Comment:
This sort of re-try logic is what is supposed to be encapsulated by `ClassUnloadCommon.triggerUnloading`, or perhaps more accurately the underlying `WhiteBox.fullGC()`. Individual tests should not have to know about such things.

Reply:
Is there a way to do something like `WhiteBox.waitForCompilationToComplete()` prior to calling `fullGC()`?

Reply:
All I see is `WhiteBox.enqueueMethodForCompilation` which will only wait for a compilation when `BackgroundCompilation` is `false`. Even then, which compilation(s) should be waited for?

We also have the issue that libgraal is not always as eager to clear object handles (see [JDK-8310218](https://bugs.openjdk.org/browse/JDK-8310218)), especially in `-Xcomp` mode and the retry loop helps in that situation as well.
If this retry logic should be in `ClassUnloadCommon.triggerUnloading`, then we'd need a second version of that method that takes a set of classes that are expected to be unloaded (e.g. `ClassUnloadCommon.triggerUnloading(String... namesOfClassesToBeUnloaded)`).

Reply:
I think we would only need that second version because the first/existing one is unreliable! But given all these "incidental" things that can keep classes strongly reachable reliability may be a pipe-dream here. But even if it is just a better version I would still prefer to see this in the utility code than individual tests.

BTW if this is a C1 issue any idea why we do not see it with C1 + C2 ?

Reply:
Because `TypeProfileWidth` is only 2 when Graal is not being used so there's very little chance of `InitExceptionUnloadTest$SpecialException` being in a type profile (see PR description).

Just to be clear, you're saying I should add `ClassUnloadCommon.triggerUnloading(String... namesOfClassesToBeUnloaded)` in this PR?

Reply:
Yes please add that and have the test use it. Thanks

Reply:
> Because TypeProfileWidth is only 2 when Graal is not being used

Okay - I have no idea what TypeProfileWidth is or does.

Reply:
> Okay - I have no idea what TypeProfileWidth is or does.

```
  product(intx, TypeProfileWidth, 2,                                        \
          "Number of receiver types to record in call/cast profile")        \
          range(0, 8) 
```

It's the (max) number of types recorded in a profile for a virtual dispatch or `instanceof` instruction.